lib/Dialect/Triton/IR/Ops.cpp

```diff
@@ -505,6 +505,57 @@ llvm::SmallVector<Type> ReduceOp::getElementTypes() {
 
 unsigned ReduceOp::getNumOperands() { return this->getOperands().size(); }
 
+namespace {
+/// Replace reduction operations with equivalent reshape operations.
+///
+/// This pattern replaces reductions whose input tensor size is 1 in the
+/// reduction dimension:
+/// ```mlir
+/// "tt.reduce"(%0, ...) <{axis = N}> ({...})
+///     : (tensor<S0 x ... x SN-1 x 1 x SN+1 x ...>, ...) ->
+///       (tensor<S0 x ... x SN-1 x SN+1 x ...>, ...)
+/// ```
+/// With equivalent reshape operations (one per operand):
+/// ```mlir
+/// tt.reshape %0 allow_reorder
+///     : tensor<S0 x ... x SN-1 x 1 x SN+1 x ...> ->
+///       tensor<S0 x ... x SN-1 x SN+1 x ...>
+/// ```
+struct CanonicalizeReshapeReducePattern final : OpRewritePattern<ReduceOp> {
+  using OpRewritePattern<ReduceOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ReduceOp reduceOp,
+                                PatternRewriter &rewriter) const final {
+    Type resultType = reduceOp->getResultTypes().front();
+    // `tensor<NxTy>->Ty` case. `tt.reshape` does not support scalar result
+    // types, so we simply skip this case.
+    if (!isa<RankedTensorType>(resultType))
+      return failure();
+    RankedTensorType inputType = reduceOp.getInputTypes().front();
+    int32_t axis = reduceOp.getAxis();
+    if (inputType.getShape()[axis] != 1)
+      return failure();
+    SmallVector<Value> reshapes(reduceOp.getNumOperands());
+    llvm::transform(
+        llvm::zip_equal(reduceOp.getSrcs(), reduceOp->getResultTypes()),
+        reshapes.begin(),
+        [loc = reduceOp.getLoc(), &rewriter](auto pair) -> Value {
+          auto &[value, resultType] = pair;
+          // Set allow_reorder to support different tensor layouts.
+          return rewriter.create<ReshapeOp>(loc, resultType, value,
+                                            /*allow_reorder=*/true);
```

Comment:
We definitely shouldn't set allow_reorder as reordering changes the result.

Reply:
This would limit applicability to `BlockedEncodingAttr` as that's the only attribute supporting non-reordering encoding. However, as we are only removing a "size 1" dimension, no reordering should take place at all. This is just a limitation of the `reshape` op IMO.

Reply:
Okay, you're welcome to remove this limitation from `ReshapeOp` if you think it's unnecessary. I don't think the right solution is to abuse `allow_reorder` like this and hope it continues to work the same in future.